Move prim ops from JIT registration to C10

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30612 Move prim ops from JIT registration to C10**

The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

Commit:
Move prim ops from JIT registration to C10

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

smessmer commented on torch/csrc/jit/register_prim_ops_c10.cpp:
```diff
@@ -0,0 +1,36 @@
+using c10::IValue;
+
+static auto registry_prim = torch::RegisterOperators().op("aten::Int.Tensor(Tensor a) -> int",
+  torch::RegisterOperators::options().kernel(c10::TensorTypeId::CPUTensorId,
```
is there a specific reason you're doing this for `CPUTensorId` only and not as a `catchAllKernel`?

iseeyuan replied:
Nope. I thought if there's tensor in argument we should use CPUTensorId. My thought can be wrong because we need to have the ops work for all kernels. I'll change it to catchAllKernel.

smessmer commented on torch/csrc/jit/register_prim_ops_c10.cpp:
```diff
@@ -0,0 +1,36 @@
+  [](at::Tensor a) -> int64_t {
+    return a.item<int64_t>();
+})
+).op(  "aten::Int.bool(bool a) -> int",
```
nit: whitespaces after the opening parentheses are weird (also below)

smessmer commented on torch/csrc/jit/register_prim_ops_c10.cpp:
```diff
@@ -0,0 +1,36 @@
+).op(  "aten::Int.bool(bool a) -> int",
+  torch::RegisterOperators::options().catchAllKernel(
+  [](bool b) -> int64_t {
+    return (int64_t)b;
```
nit: `static_cast` is better than C-style casts. This was already wrong in the old code, but let's fix it since we touch this code anyways.

smessmer reviewed (approved):
Looks good. I had to revert my PR fixing the boxed variable dispatch, so this will currently likely fail CI. I'll try to re-land it as soon as I can and will let you know when you can land this.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

zdevito commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -38,7 +38,9 @@ void Function::append_operator(const std::string& name,
   // Add "_" prefix to work around the double registration both of jit/generated
   // and here. TODO: remove it when we have separate build for lite interpreter.
-  opname.name = "_" + opname.name;
+  if (opname.name != "aten::Int") {
```
this is very strange. certainly not ok without a comment, probably not ok in general.

iseeyuan replied:
Yeah it does look strange. It may not be possible to move all Aten ops to C10 in one shot, so the plan is to move them incrementally. I'll add a comment there. Hopefully this moving process will not last long so that the if statement with comment will be there temporarily.

zdevito commented on torch/csrc/jit/register_prim_ops_c10.cpp:
```diff
@@ -0,0 +1,36 @@
+using c10::IValue;
+
+static auto registry_prim = torch::RegisterOperators().op("aten::Int.Tensor(Tensor a) -> int",
+  torch::RegisterOperators::options().catchAllKernel(
```
`torch::RegisterOperators::options().catchAllKernel` <- this boilerplate is really big. Can we make this shorter? There are a log of registrations in register_prim_ops and having to read around it will be difficuilt.

iseeyuan replied:
@smessmer Could we make it shorter?

smessmer replied:
cc @charliechen0401 on if we can already skip the `aliasAnalysis()` call here. If we can skip the `aliasAnalysis()` call, then we can use the shorthand API:

    static auto registry_prim = torch::RegisterOperators().op("aten::Int.Tensor(Tensor a) -> int", [] (at::Tensor a) -> int64_t {...}))

smessmer replied:
oh wait those ops aren't specified in `native_functions.yaml`. Then, we actually do need to specify `aliasAnalysis()` and the shorthand doesn't work.

iseeyuan replied:
Yep, aliasAnalysis() is required here. Otherwise there will be runtime error complaining on the in-consistent input shape.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

kostmo commented:
## CircleCI build failures summary
As of commit 62d9ceb:
* **11/11** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=62d9ceb0413d05b444de99310e04042aab8239ee).

### 11 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965075) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (1/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75184151))
<details>
<summary>
<code>Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody'</code>
</summary>

```
Dec 16 17:10:05 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/autograd.cpp.o 
Dec 16 17:10:05 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/any.cpp.o 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'JitTest_AutogradSymbols_Test' 
Dec 16 17:10:05 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:10:05                 ^ 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'test_info_' 
Dec 16 17:10:05 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:10:05                 ^ 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody' 
Dec 16 17:10:05 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:10:05                 ^ 
Dec 16 17:10:05 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:10:05 3 errors generated. 
Dec 16 17:10:05 make[2]: *** [test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o] Error 1 
Dec 16 17:10:05 test_jit/CMakeFiles/test_jit.dir/build.make:86: recipe for target 'test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o' failed 
Dec 16 17:10:05 CMakeFiles/Makefile2:11239: recipe for target 'test_jit/CMakeFiles/test_jit.dir/all' failed 
Dec 16 17:10:05 make[1]: *** [test_jit/CMakeFiles/test_jit.dir/all] Error 2 
Dec 16 17:10:05 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 17:10:05 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/dataloader.cpp.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965060) caffe2_onnx_py2_gcc5_ubuntu16_04_build (2/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75185711))
<details>
<summary>
<code>Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()'</code>
</summary>

```
Dec 16 17:10:54  ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'testing::TestInfo* const torch::jit::JitTest_AutogradSymbols_Test::test_info_' previously declared here 
Dec 16 17:10:54    TEST(JitTest, name) { \ 
Dec 16 17:10:54         ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:78:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:10:54    _(AutogradSymbols)                   \ 
Dec 16 17:10:54    ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:10:54  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:10:54  ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' 
Dec 16 17:10:54    TEST(JitTest, name) { \ 
Dec 16 17:10:54         ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:80:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:10:54    _(AutogradSymbols)                   \ 
Dec 16 17:10:54    ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:10:54  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:10:54  ^ 
Dec 16 17:10:54 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'virtual void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' previously defined here 
Dec 16 17:10:54    TEST(JitTest, name) { \ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965074) pytorch_xla_linux_xenial_py3_6_clang7_build (3/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75186253))
<details>
<summary>
<code>Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody'</code>
</summary>

```
Dec 16 17:11:55 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/any.cpp.o 
Dec 16 17:11:56 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/dataloader.cpp.o 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'JitTest_AutogradSymbols_Test' 
Dec 16 17:11:56 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:11:56                 ^ 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'test_info_' 
Dec 16 17:11:56 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:11:56                 ^ 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody' 
Dec 16 17:11:56 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:11:56                 ^ 
Dec 16 17:11:56 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:11:56 3 errors generated. 
Dec 16 17:11:56 make[2]: *** [test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o] Error 1 
Dec 16 17:11:56 test_jit/CMakeFiles/test_jit.dir/build.make:86: recipe for target 'test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o' failed 
Dec 16 17:11:56 make[1]: *** [test_jit/CMakeFiles/test_jit.dir/all] Error 2 
Dec 16 17:11:56 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 17:11:56 CMakeFiles/Makefile2:11210: recipe for target 'test_jit/CMakeFiles/test_jit.dir/all' failed 
Dec 16 17:11:56 [ 91%] Building CXX object test_api/CMakeFiles/test_api.dir/enum.cpp.o 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965064) pytorch_linux_xenial_py3_clang5_asan_build (4/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75186370))
<details>
<summary>
<code>Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody'</code>
</summary>

```
Dec 16 17:12:35 [ 92%] Built target test_cpp_rpc 
Dec 16 17:12:35 [ 92%] Building CXX object test_jit/CMakeFiles/test_jit.dir/test_ivalue.cpp.o 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'JitTest_AutogradSymbols_Test' 
Dec 16 17:12:35 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:12:35                 ^ 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'test_info_' 
Dec 16 17:12:35 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:12:35                 ^ 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody' 
Dec 16 17:12:35 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:12:35                 ^ 
Dec 16 17:12:35 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 17:12:35 3 errors generated. 
Dec 16 17:12:35 test_jit/CMakeFiles/test_jit.dir/build.make:86: recipe for target 'test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o' failed 
Dec 16 17:12:35 make[2]: *** [test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o] Error 1 
Dec 16 17:12:35 make[2]: *** Waiting for unfinished jobs.... 
Dec 16 17:12:35 make[1]: *** [test_jit/CMakeFiles/test_jit.dir/all] Error 2 
Dec 16 17:12:35 make[1]: *** Waiting for unfinished jobs.... 
Dec 16 17:12:35 CMakeFiles/Makefile2:10987: recipe for target 'test_jit/CMakeFiles/test_jit.dir/all' failed 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965073) pytorch_linux_xenial_py2_7_9_build (5/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75186410))
<details>
<summary>
<code>Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()'</code>
</summary>

```
Dec 16 17:12:45  ^~~~~~~~~~~~~~~ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'testing::TestInfo* const torch::jit::JitTest_AutogradSymbols_Test::test_info_' previously declared here 
Dec 16 17:12:45    TEST(JitTest, name) { \ 
Dec 16 17:12:45         ^ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:78:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:12:45    _(AutogradSymbols)                   \ 
Dec 16 17:12:45    ^ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:12:45  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:12:45  ^~~~~~~~~~~~~~~ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' 
Dec 16 17:12:45    TEST(JitTest, name) { \ 
Dec 16 17:12:45         ^ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:80:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:12:45    _(AutogradSymbols)                   \ 
Dec 16 17:12:45    ^ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:12:45  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:12:45  ^~~~~~~~~~~~~~~ 
Dec 16 17:12:45 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'virtual void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' previously defined here 
Dec 16 17:12:45    TEST(JitTest, name) { \ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965063) pytorch_macos_10_13_py3_build (6/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75186523))
<details>
<summary>
<code>Dec 16 09:13:22 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o</code>
</summary>

```
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx.a(common_avx.cc.o) has no symbols 
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx.a(common_avx.cc.o) has no symbols 
Dec 16 09:13:20 [2631/2973] Linking CXX static library lib/libCaffe2_perfkernels_avx2.a 
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx2.a(common_avx2.cc.o) has no symbols 
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx2.a(common_avx2.cc.o) has no symbols 
Dec 16 09:13:20 [2632/2973] Linking CXX static library lib/libCaffe2_perfkernels_avx512.a 
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx512.a(common_avx512.cc.o) has no symbols 
Dec 16 09:13:20 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx512.a(common_avx512.cc.o) has no symbols 
Dec 16 09:13:20 warning: /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: warning for library: lib/libCaffe2_perfkernels_avx512.a the table of contents is empty (no object file members in the library define global symbols) 
Dec 16 09:13:22 [2633/2973] Building CXX object test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o 
Dec 16 09:13:22 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o  
DKs/MacOSX10.13.sdk -mmacosx-version-min=10.9 -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -MF test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o.d -o test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -c ../test/cpp/jit/gtest.cpp 
Dec 16 09:13:22 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'JitTest_AutogradSymbols_Test' 
Dec 16 09:13:22 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 09:13:22                 ^ 
Dec 16 09:13:22 ../test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 09:13:22 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'test_info_' 
Dec 16 09:13:22 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 09:13:22                 ^ 
Dec 16 09:13:22 ../test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 09:13:22 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965055) pytorch_linux_xenial_py3_6_gcc5_4_build (7/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75186725))
<details>
<summary>
<code>Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()'</code>
</summary>

```
Dec 16 17:14:34  ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'testing::TestInfo* const torch::jit::JitTest_AutogradSymbols_Test::test_info_' previously declared here 
Dec 16 17:14:34    TEST(JitTest, name) { \ 
Dec 16 17:14:34         ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:78:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:14:34    _(AutogradSymbols)                   \ 
Dec 16 17:14:34    ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:14:34  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:14:34  ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: error: redefinition of 'void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' 
Dec 16 17:14:34    TEST(JitTest, name) { \ 
Dec 16 17:14:34         ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/tests.h:80:3: note: in expansion of macro 'JIT_GTEST' 
Dec 16 17:14:34    _(AutogradSymbols)                   \ 
Dec 16 17:14:34    ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro 'TH_FORALL_TESTS' 
Dec 16 17:14:34  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:14:34  ^ 
Dec 16 17:14:34 /var/lib/jenkins/workspace/test/cpp/jit/gtest.cpp:9:8: note: 'virtual void torch::jit::JitTest_AutogradSymbols_Test::TestBody()' previously defined here 
Dec 16 17:14:34    TEST(JitTest, name) { \ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965062) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (8/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75187903))
<details>
<summary>
<code>Dec 16 17:17:04 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o</code>
</summary>

```
Dec 16 17:17:02 [2739/3000] Linking CXX executable bin/SubgraphMatcherTest 
Dec 16 17:17:02 [2740/3000] Building CXX object caffe2/CMakeFiles/cast_test.dir/utils/cast_test.cc.o 
Dec 16 17:17:03 [2741/3000] Building CXX object caffe2/CMakeFiles/MatchTest.dir/core/nomnigraph/tests/MatchTest.cc.o 
Dec 16 17:17:03 [2742/3000] Linking CXX executable bin/cast_test 
Dec 16 17:17:03 [2743/3000] Linking CXX executable bin/MatchTest 
Dec 16 17:17:03 [2744/3000] Linking CXX executable bin/generate_proposals_op_test 
Dec 16 17:17:03 [2745/3000] Building CXX object caffe2/CMakeFiles/distributed_test.dir/opt/distributed_test.cc.o 
Dec 16 17:17:04 [2746/3000] Linking CXX executable bin/distributed_test 
Dec 16 17:17:04 [2747/3000] Building CXX object test_jit/CMakeFiles/test_jit.dir/__/common/main.cpp.o 
Dec 16 17:17:04 [2748/3000] Building CXX object test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o 
Dec 16 17:17:04 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o  
-errno -fno-trapping-math -DHAVE_AVX_CPU_DEFINITION -DHAVE_AVX2_CPU_DEFINITION -O3  -fPIE   -DCAFFE2_USE_GLOO -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -fopenmp -std=gnu++14 -MD -MT test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -MF test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o.d -o test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -c ../test/cpp/jit/gtest.cpp 
Dec 16 17:17:04 In file included from /pytorch/third_party/googletest/googletest/include/gtest/gtest.h:59:0, 
Dec 16 17:17:04                  from ../test/cpp/jit/gtest.cpp:3: 
Dec 16 17:17:04 ../test/cpp/jit/gtest.cpp:9:8: error: redefinition of ‘class torch::jit::JitTest_AutogradSymbols_Test’ 
Dec 16 17:17:04    TEST(JitTest, name) { \ 
Dec 16 17:17:04         ^ 
Dec 16 17:17:04 ../test/cpp/jit/tests.h:80:3: note: in expansion of macro ‘JIT_GTEST’ 
Dec 16 17:17:04    _(AutogradSymbols)                   \ 
Dec 16 17:17:04    ^ 
Dec 16 17:17:04 ../test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro ‘TH_FORALL_TESTS’ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965068) pytorch_macos_10_13_cuda10_0_cudnn7_py3_build (9/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75188101))
<details>
<summary>
<code>Dec 16 09:18:04 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o</code>
</summary>

```
Dec 16 09:18:02 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx.a(common_avx.cc.o) has no symbols 
Dec 16 09:18:02 [2630/2973] Building CXX object test_jit/CMakeFiles/test_jit.dir/__/common/main.cpp.o 
Dec 16 09:18:02 [2631/2973] Linking CXX static library lib/libCaffe2_perfkernels_avx2.a 
Dec 16 09:18:02 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx2.a(common_avx2.cc.o) has no symbols 
Dec 16 09:18:02 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx2.a(common_avx2.cc.o) has no symbols 
Dec 16 09:18:02 [2632/2973] Linking CXX static library lib/libCaffe2_perfkernels_avx512.a 
Dec 16 09:18:02 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx512.a(common_avx512.cc.o) has no symbols 
Dec 16 09:18:02 /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: lib/libCaffe2_perfkernels_avx512.a(common_avx512.cc.o) has no symbols 
Dec 16 09:18:02 warning: /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: warning for library: lib/libCaffe2_perfkernels_avx512.a the table of contents is empty (no object file members in the library define global symbols) 
Dec 16 09:18:04 [2633/2973] Building CXX object test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o 
Dec 16 09:18:04 FAILED: test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o  
k -mmacosx-version-min=10.9 -fPIE   -DCAFFE2_USE_GLOO -DCUDA_HAS_FP16=1 -DHAVE_GCC_GET_CPUID -DUSE_AVX -DUSE_AVX2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces "-Xpreprocessor -fopenmp -I/usr/local/include" -std=gnu++14 -MD -MT test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -MF test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o.d -o test_jit/CMakeFiles/test_jit.dir/gtest.cpp.o -c ../test/cpp/jit/gtest.cpp 
Dec 16 09:18:04 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'JitTest_AutogradSymbols_Test' 
Dec 16 09:18:04 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 09:18:04                 ^ 
Dec 16 09:18:04 ../test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 09:18:04 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'test_info_' 
Dec 16 09:18:04 TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 09:18:04                 ^ 
Dec 16 09:18:04 ../test/cpp/jit/gtest.cpp:12:17: note: previous definition is here 
Dec 16 09:18:04 ../test/cpp/jit/gtest.cpp:12:17: error: redefinition of 'TestBody' 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965059) binary_linux_manywheel_2_7mu_cpu_devtoolset7_build (10/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75188543))
<details>
<summary>
<code>Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:9:8: error: redefinition of ‘void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’</code>
</summary>

```
Dec 16 17:18:14  ^~~~~~~~~~~~~~~ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:9:8: note: ‘testing::TestInfo* const torch::jit::JitTest_AutogradSymbols_Test::test_info_’ previously declared here 
Dec 16 17:18:14    TEST(JitTest, name) { \ 
Dec 16 17:18:14         ^ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/tests.h:78:3: note: in expansion of macro ‘JIT_GTEST’ 
Dec 16 17:18:14    _(AutogradSymbols)                   \ 
Dec 16 17:18:14    ^ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro ‘TH_FORALL_TESTS’ 
Dec 16 17:18:14  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:18:14  ^~~~~~~~~~~~~~~ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:9:8: error: redefinition of ‘void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’ 
Dec 16 17:18:14    TEST(JitTest, name) { \ 
Dec 16 17:18:14         ^ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/tests.h:80:3: note: in expansion of macro ‘JIT_GTEST’ 
Dec 16 17:18:14    _(AutogradSymbols)                   \ 
Dec 16 17:18:14    ^ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro ‘TH_FORALL_TESTS’ 
Dec 16 17:18:14  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:18:14  ^~~~~~~~~~~~~~~ 
Dec 16 17:18:14 /pytorch/test/cpp/jit/gtest.cpp:9:8: note: ‘virtual void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’ previously defined here 
Dec 16 17:18:14    TEST(JitTest, name) { \ 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3965076) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build (11/11)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=75188751))
<details>
<summary>
<code>Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:9:8: error: redefinition of ‘void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’</code>
</summary>

```
Dec 16 17:19:22  ^~~~~~~~~~~~~~~ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:9:8: note: ‘testing::TestInfo* const torch::jit::JitTest_AutogradSymbols_Test::test_info_’ previously declared here 
Dec 16 17:19:22    TEST(JitTest, name) { \ 
Dec 16 17:19:22         ^ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/tests.h:78:3: note: in expansion of macro ‘JIT_GTEST’ 
Dec 16 17:19:22    _(AutogradSymbols)                   \ 
Dec 16 17:19:22    ^ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro ‘TH_FORALL_TESTS’ 
Dec 16 17:19:22  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:19:22  ^~~~~~~~~~~~~~~ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:9:8: error: redefinition of ‘void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’ 
Dec 16 17:19:22    TEST(JitTest, name) { \ 
Dec 16 17:19:22         ^ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/tests.h:80:3: note: in expansion of macro ‘JIT_GTEST’ 
Dec 16 17:19:22    _(AutogradSymbols)                   \ 
Dec 16 17:19:22    ^ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:12:1: note: in expansion of macro ‘TH_FORALL_TESTS’ 
Dec 16 17:19:22  TH_FORALL_TESTS(JIT_GTEST) 
Dec 16 17:19:22  ^~~~~~~~~~~~~~~ 
Dec 16 17:19:22 /pytorch/test/cpp/jit/gtest.cpp:9:8: note: ‘virtual void torch::jit::JitTest_AutogradSymbols_Test::TestBody()’ previously defined here 
Dec 16 17:19:22    TEST(JitTest, name) { \ 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

iseeyuan commented:
Keep the change in one file to track the updates.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

iseeyuan commented:
There are operators that will not appear in mobile. For example, prim::profile and prim::FusionGroup. They depend on JIT node or graph. The c10 ops without those dependencies may stay in a separate file, which can be used to build mobile.

Commit:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

iseeyuan committed:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

[ghstack-poisoned]

iseeyuan committed:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

[ghstack-poisoned]

iseeyuan committed:
Update on "Move prim ops from JIT registration to C10"


The first version to move prim ops to c10 registration. After the reviewers are fine with the initial changes, more operators will be moved in the same style.

Differential Revision: [D19237648](https://our.internmc.facebook.com/intern/diff/D19237648)

[ghstack-poisoned]
